server/server/src/billing/billing_service.rs

```diff
@@ -1,194 +1,361 @@
-                "User is already in payment flow, or this request is too soon after a failed one. public_key: {}",
-                keys::stringify_public_key(public_key)
-            );
-            return Err(Abort(ClientError(SwitchAccountTierError::ConcurrentRequestsAreTooSoon)));
         }
+    }
+
+    reset_billing_lock(&context.public_key, &mut con, &mut sub_profile).await?;
 
-        user_info.last_in_payment_flow = current_time;
+    Ok(CancelSubscriptionResponse {})
+}
+
+pub async fn upgrade_account_stripe(
+    context: RequestContext<'_, UpgradeAccountStripeRequest>,
+) -> Result<UpgradeAccountStripeResponse, ServerError<UpgradeAccountStripeError>> {
+    let (request, server_state) = (&context.request, context.server_state);
 
-        pipe.json_set(stripe_user_info(public_key), &user_info)?;
+    info!("Attempting to switch account tier of {:?} to premium", context.public_key);
 
-        Ok(&mut pipe)
+    let mut con = server_state.index_db_pool.get().await?;
+
+    let mut sub_profile = lock_billing_workflow(
+        &context.public_key,
+        &mut con,
+        server_state.config.stripe.millis_between_user_payment_flows,
+    )
+    .await
+    .map_err(|err| match err {
+        ClientError(LockBillingWorkflowError::ExistingRequestPending) => {
+            ClientError(UpgradeAccountStripeError::ExistingRequestPending)
+        }
+        InternalError(msg) => InternalError(msg),
+    })?;
+
+    if sub_profile.data_cap == PREMIUM_TIER_USAGE_SIZE {
+        return Err(ClientError(UpgradeAccountStripeError::AlreadyPremium));
+    }
+
+    let maybe_user_info = sub_profile.billing_platform.and_then(|info| {
+        if let BillingPlatform::Stripe(stripe_info) = info {
+            Some(stripe_info)
+        } else {
+            None
+        }
     });
-    return_if_error!(tx_result);
+
+    let (user_info, new_data_cap) = create_subscription(
+        server_state,
+        &mut con,
+        &context.public_key,
+        &request.account_tier,
+        maybe_user_info,
+    )
+    .await?;
+
+    sub_profile.billing_platform = Some(BillingPlatform::Stripe(user_info));
+    sub_profile.data_cap = new_data_cap;
+
+    reset_billing_lock(&context.public_key, &mut con, &mut sub_profile).await?;
 
     info!(
-        "User successfully entered payment flow. public_key: {}",
-        keys::stringify_public_key(public_key)
+        "Successfully switched the account tier of {:?} from free to premium.",
+        context.public_key
     );
 
-    Ok(user_info)
+    Ok(UpgradeAccountStripeResponse {})
 }
 
 async fn create_subscription(
     server_state: &ServerState, con: &mut deadpool_redis::Connection, public_key: &PublicKey,
-    fmt_public_key: &str, payment_method: &PaymentMethod, user_info: &mut StripeUserInfo,
-) -> Result<u64, ServerError<SwitchAccountTierError>> {
-    let (customer_id, payment_method_id) = match payment_method {
+    account_tier: &StripeAccountTier, maybe_user_info: Option<StripeUserInfo>,
```

Comment:
why did this become optional?

Reply:
I made this optional since I wanted `StripeUserInfo` to not be initialized when there was no information. Back then I passed a `StripeUserInfo` into this function regardless of whether it had any relevant info (like a past credit card), but this then required me to have an Option for customer_id and other related fields. I rather `StripeUserInfo` and `GooglePlayUserInfo` definitely contain information about the user it's talking about, and not have to deal with `Option`s in their fields.